### ISSUE TITLE
Deprecate th:include

### ISSUE THREAD

**Issue by danielfernandez** [thymeleaf/thymeleaf#405]
Once `th:insert` has been introduced in Thymeleaf 3.0 and documentation for `th:include` has been retired, the latter should be deprecated (with warnings being issued in Thymeleaf 3.1).

This `th:include` should be removed in 3.2/4.0.

**Comment by Antibrumm** [comment]
Could you please share the reason why include should be retired? In my project we use include quite often as it allows us to create fragments that show a correct prototype and also allow to build the inclusion side according to the same style without the need for tag removal. Or fragments that are included in a template but also rerendered via ajax. I will share one template as soon as I got some time so I hope I can show why I think include still has a good reason to stick around.

**Comment by danielfernandez** [comment]
Hi @Antibrumm. The reason is that the way it worked caused a good amount of confusion among many users: *template inclusion* operations in other template systems usually include entire templates and, though there aren't many other options out there for comparison that are able to include only *fragments* of other templates, many people assumed `th:include` would include the host tag too, just like `th:insert` does now.

So that's why we introduced `th:insert`, which is an type of inclusion we didn't have before, and decided to deprecate `th:include` for 3.1. Having both, plus `th:replace`, would be too many options for what actually is almost the same operation, only with small operational differences. 

Besides, `th:replace`'s nature is precisely to avoid the need for tag removal during tag inclusion by directly bringing the host tag to the includer, so many scenarios can be *updated* with a simple switch from `th:include` to `th:replace`. In way, `th:include` and `th:replace` were doing the same kind of fragment insertion, whereas `th:insert` gives us something new that feels quite natural to many people coming from other template systems.

And as a last point, though this is a mere *implementation detail*, the logic for implementing `th:include` in an event-based engine like Thymeleaf is since v3 is actually considerably more complicated than `th:insert` and `th:replace`. Even if this would not be a significative difference for most people, the fact is that `th:insert`- and `th:replace`-based *fragment expressions* should perform a bit better than `th:include`.

**Comment by Antibrumm** [comment]
Thanks Daniel for the explanation. I see your point on the confusion part. I also needed to write some simple tests to get the differences between the include modes. 

The main advantage that `include` had over `replace` was that you were still able to declare variables on the same element as the "design element" was replaced after evaluation. This made the "toplevel-templates" pretty slick :)

    --- Main Template
    <ul class="simpleEditor">
        <li th:include="::text" with:text="Changed 1" .../>
        <li th:include="::input" with:attribute="someAttr" .../>
        <li th:include="::datePicker" with:attribute="someDateAttr" .../>
    </ul>

    --- two component fragments displayed in the same style as the final editor
    <ul class="simpleEditor">
        <li th:fragment="text"><label>label</label><span th:text="text">Some Text</span></li>
        <li th:fragment="input"><label>label</label><input th:field="${attribute}"/></li>
    </ul>

Actually we wrote a whole bunch of form input fragments.

Just switching to `insert` will now produce duplicate `<li></li>` tags. Using `th:block` would be an option but then the "natural templating" does not work anymore.

That would mean we have two options for that case:
 - Use `th:insert="..." th:remove="tag"` (haven't seen a big performance impact on that, `th:block` might be better?)
 - Or, I think this can now be solved by using `fragment parameters.` Using parameters we should be able to use `replace` as the fragment expression has to be evaluated before. Gonna check that.
